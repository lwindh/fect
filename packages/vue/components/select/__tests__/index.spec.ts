import { mount } from '@vue/test-utils'
import Select from '..'
import Option from '../../select-option'

const Wrapper = {
  components: {
    [Select.name]: Select,
    [Option.name]: Option
  },
  data() {
    return {
      val: '',
      placeholder: '摸鱼是肯定的要摸鱼的',
      multiple: false,
      clearable: true,
      disabled: false,
      size: 'medium',
      childrenDisabled: false
    }
  },
  template: `
    <div class="container">
     <fe-select 
     ref="selectRef"
     v-model="val"
     :multiple="multiple"
     :placeholder="placeholder"
     :disabled="disabled"
     :clearable="clearable"
     :size="size"
     :data-size="size"
     >
       <fe-option label="吃饭" value="1"></fe-option>
       <fe-option label="摸鱼" value="2" :disabled="childrenDisabled"></fe-option>
     </fe-select>
    </div>
    `
}

describe('Select', () => {
  it('should be render as a element', async () => {
    const wrapper = mount(Wrapper, { attachTo: document.body })
    const {
      selectRef: { setTeleport }
    } = wrapper.vm.$refs as any
    await setTeleport('.container')
    expect(wrapper.html()).toMatchSnapshot()
    expect(() => wrapper.unmount()).not.toThrow()
  })
  it('component props should be work correctly', async () => {
    const wrapper = mount(Wrapper, { attachTo: document.body })
    const {
      selectRef: { setTeleport }
    } = wrapper.vm.$refs as any
    await setTeleport('.container')
    const selectEl = wrapper.find('.fect-select')
    await selectEl.trigger('click')
    await wrapper.setData({ size: 'mini', disabled: true })
    expect(selectEl.attributes('data-size')).toBe('mini')
    await wrapper.setData({ size: 'xxx', disabled: false })
    expect(selectEl.attributes('style')).toContain('--select-fontSize: 14px;')
    await wrapper.setData({ val: ['1', '2'], multiple: true })
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })

  it('clean icon should be work', async () => {
    const wrapper = mount(Wrapper, { attachTo: document.body })
    const {
      selectRef: { setTeleport }
    } = wrapper.vm.$refs as any
    await setTeleport('.container')
    const optionEls = wrapper.findAll('.fect-option')
    await optionEls[0].trigger('click')
    const selectInputEl = wrapper.find('.value')
    expect(selectInputEl.exists()).toBeTruthy()
    expect(selectInputEl.text()).toBe('吃饭')
    const selectEl = wrapper.find('.fect-select')
    await selectEl.trigger('mouseenter')
    await selectEl.trigger('mouseleave')
    await selectEl.trigger('mouseenter')
    const el = wrapper.find('.fect-select__clearIcon')
    await el.trigger('click')
    await wrapper.setData({ multiple: true, val: [] })
    await optionEls[0].trigger('click')
    await optionEls[1].trigger('click')
    const multipleEls = wrapper.findAll('.fect-select__clearIcon')
    await multipleEls[1].trigger('click')
    expect(wrapper.vm.val).toEqual(['1'])
    await wrapper.setData({ childrenDisabled: true })
    await optionEls[1].trigger('click')
    await wrapper.setData({ disabled: true })
    await selectEl.trigger('click')
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })

  it('useClickAway should be work', async () => {
    const wrapper = mount(Wrapper, { attachTo: document.body })
    await wrapper.setData({ multiple: true, val: [] })
    const {
      selectRef: { setTeleport }
    } = wrapper.vm.$refs as any
    await setTeleport('.container')
    const selectEl = wrapper.find('.fect-select')
    const inputEl = wrapper.find('input')
    await selectEl.trigger('click')
    expect(inputEl.attributes('aria-expanded')).toBe('true')
    const containerEl = wrapper.find('.container')
    await containerEl.trigger('click')
    expect(inputEl.attributes('aria-expanded')).toBe('false')
    wrapper.unmount()
  })
  it('when in multiply , initial value as empty should render placeholder', () => {
    const wrapper = mount(Select, {
      props: {
        multiple: true,
        modelValue: []
      }
    })
    expect(wrapper.find('.fect-select__placeholder').exists()).toBeTruthy()
  })
})
