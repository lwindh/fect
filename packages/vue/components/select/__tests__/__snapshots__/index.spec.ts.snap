// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component props should be work correctly 1`] = `
<div class="container">
  <div class="fect-select multiple" style="--select-minWidth: 160px; --select-height: calc(1.688 * var(--fay-gap)); --select-fontSize: 14px;" data-size="xxx">
    <div class="fect-grid__group fect-grid__group--wrap " style="align-content: flex-start; align-items: stretch; flex-direction: row; justify-content: flex-start; --fect-grid-gap: calc(0.5 * 16px * 1/3);">
      <div class="fect-grid " style="align-content: flex-start; align-items: stretch; flex-direction: row; justify-content: flex-start;">
        <div class="fect-select__item">吃饭<div class="fect-select__clearIcon"><svg viewBox="0 0 24 24" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" fill="none" shape-rendering="geometricPrecision" color="currentColor">
              <path d="M18 6L6 18"></path>
              <path d="M6 6l12 12"></path>
            </svg></div>
        </div>
      </div>
      <div class="fect-grid " style="align-content: flex-start; align-items: stretch; flex-direction: row; justify-content: flex-start;">
        <div class="fect-select__item">摸鱼<div class="fect-select__clearIcon"><svg viewBox="0 0 24 24" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" fill="none" shape-rendering="geometricPrecision" color="currentColor">
              <path d="M18 6L6 18"></path>
              <path d="M6 6l12 12"></path>
            </svg></div>
        </div>
      </div>
    </div>
    <!--teleport start-->
    <!--teleport end-->
    <div class="fect-select__icon-context click"><svg viewBox="0 0 24 24" width="20px" height="20px" stroke-width="1" stroke-linecap="round" color="currentColor" stroke="currentColor" stroke-linejoin="round" fill="none" shape-rendering="geometricPrecision">
        <path d="M6 9l6 6 6-6"></path>
      </svg></div>
  </div>
  <div class="fect-select__dropdown" style="width: 0px; top: 2px; left: 0px;">
    <div class="fect-option ">吃饭</div>
    <div class="fect-option ">摸鱼</div>
  </div>
</div>
`;

exports[`Select should be render as a element 1`] = `
<div class="container">
  <div class="fect-select" style="--select-minWidth: 160px; --select-height: calc(1.688 * var(--fay-gap)); --select-fontSize: 14px;" data-size="medium"><span class="value fect-select__placeholder">摸鱼是肯定的要摸鱼的</span>
    <!--teleport start-->
    <!--teleport end-->
    <div class="fect-select__icon-context"><svg viewBox="0 0 24 24" width="20px" height="20px" stroke-width="1" stroke-linecap="round" color="currentColor" stroke="currentColor" stroke-linejoin="round" fill="none" shape-rendering="geometricPrecision">
        <path d="M6 9l6 6 6-6"></path>
      </svg></div>
  </div>
  <div class="fect-select__dropdown" style="display: none;">
    <div class="fect-option ">吃饭</div>
    <div class="fect-option ">摸鱼</div>
  </div>
</div>
`;

exports[`Select should change value when trigger a value 1`] = `
<div class="container">
  <div class="fect-select" style="--select-minWidth: 160px; --select-height: calc(1.688 * var(--fay-gap)); --select-fontSize: 14px;" data-size="medium"><span class="value">吃饭</span>
    <!--teleport start-->
    <!--teleport end-->
    <div class="fect-select__icon-context"><svg viewBox="0 0 24 24" width="20px" height="20px" stroke-width="1" stroke-linecap="round" color="currentColor" stroke="currentColor" stroke-linejoin="round" fill="none" shape-rendering="geometricPrecision">
        <path d="M6 9l6 6 6-6"></path>
      </svg></div>
  </div>
  <div class="fect-select__dropdown" style="display: none;">
    <div class="fect-option ">吃饭</div>
    <div class="fect-option disabled">摸鱼</div>
  </div>
</div>
`;
